### ISSUE TITLE
[Feature] Manual cell counting in brainreg-segment

### ISSUE THREAD

**Issue by janfpl** [brainglobe/cellfinder#160]
I have whole mouse brain datasets with relatively sparse labeling. Rather than using the fully automated method of segmenting and classifying cells, I would like to be able to count them manually and register cell positions to atlas space for visualization and quantification. 

I would like to use the brainreg-segment GUI to mark cell positions, and export a csv with the manually annotated cell positions. 

Ideally, there would be multiple counters such as the one found in ImageJ. Three counters would allow for marking say, GFP+ cells, mCherry+ cells, and double-labeled cells.

**Comment by adamltyson** [comment]
Thanks for raising the issue. This is something I've wanted to support for a while, but I've yet to find the time do it properly.

I've got part way to making this, and it's available in cellfinder now (but not advertised or documented). If you install a new version of cellfinder (version 0.4.8 will be released in the next 24 hours), you can use the tool `cellfinder_curate_new`. If you want to give this a go, and let me know what feedback you have, that would be great.

**Basic instructions:**
* Run `cellfinder_curate_new` from the command line, and a napari window should open
* First load the raw data, by clicking Load signal. This is the whole directory of the your images (e.g. as 2D tiffs) . You should be able to navigate around the whole stack, change contrast limits, gamma etc. This should all be pretty straightforward. Just don't click the buttons in the bottom left, many of those will end up trying to load the whole dataset into ram
* Start the counting with  Add cell count
* Make sure the cells layer on the left is highlighted, then click the + at the top to start adding cells.
* Click on cells, and a 3D sphere will appear around them, hopefully highlighting them in 3D
* You can switch to a pan and zoom mode with Z and back to adding cells with P. Help -> Show key bindings will show any other keyboard shortcuts that are available
* Once you've selected some cells, you can save them with Save cells. Choose a directory (or make a new one), then it will save a cells.xml file
* Then you can analyse the cell positions with Analyse cells. This will ask for a brainreg output directory, (this will need to be run separately on the same input data).
* It will then run for a while, and then you'll see the various files, like a summary of all cells, and a csv of each cell position.

This tool is nowhere near ready for release, but maybe it's useful for you? I'd also be interested to know if you think it should be it's own tool as part of cellfinder, or integrated as a third data type in `brainreg-segment` (in addition to volumes and lines).

Hope this helps in some way!

**Comment by janfpl** [comment]
I followed your instructions and it worked well! Thank you for your help. I will spend some time quantifying my data and will provide some more feedback.

A few things that have come up so far:
1. After saving some cells to csv, and reopening the cells with the 'Load cell count' to add more, I can change the symbol of the initially labeled cells, but not the color. They remain white despite changing both the 'face color' or 'edge color' properties. Adding additional cell markers (ones not included in the first count) do show the updated color properties.

2. One issue I'm finding is that cells at the border of one region are being incorrectly registered in an adjacent region or fiber tract. It would be helpful if there was a way to overlay the boundaries of the registered atlas during the cell marking phase, so that you could mark the corresponding point in the atlas accurately. Another helpful feature would be to display the region corresponding to where the cursor position was (I think this worked with amap?).

**Comment by adamltyson** [comment]
Thanks for the feedback, and thanks for your patience. This tool was made quickly for a user in my lab to do some quick analyses for a paper, so it's a long way off being finished.

> After saving some cells to csv, and reopening the cells with the 'Load cell count' to add more, I can change the symbol of the initially labeled cells, but not the color. They remain white despite changing both the 'face color' or 'edge color' properties. Adding additional cell markers (ones not included in the first count) do show the updated color properties.

This tool is built upon napari, and I think this is an issue on their side (https://github.com/brainglobe/cellfinder/issues/103, https://github.com/napari/napari/issues/1760). I think I can add a button to add this functionality though.

> One issue I'm finding is that cells at the border of one region are being incorrectly registered in an adjacent region or fiber tract. It would be helpful if there was a way to overlay the boundaries of the registered atlas during the cell marking phase, so that you could mark the corresponding point in the atlas accurately. Another helpful feature would be to display the region corresponding to where the cursor position was (I think this worked with amap?).

This is definitely possible. It's implemented within [brainreg-segment](https://github.com/brainglobe/brainreg-segment), and I can copy the code over to this tool.  There's a scientific consideration of whether highlighting cells with their registered atlas positions known would introduce bias, but I'll add this and let the users decide how to use it.

I'm afraid it will be a while before I find the time to work any more on this, but your feedback is really helpful. If you or anyone else wants to contribute though, a lot of the work is already done, and I can provide guidance.

**Comment by janfpl** [comment]
While introducing bias is a valid concern, I'm trying to correct for a few pixels of error in the registration that is making the difference between a cell being registered in the corpus callosum instead of hippocampal CA3. If you have any ideas on a solution it'd be much appreciated.

I tried taking the downsampled boundaries.tif from the brainreg directory and rescaling it to the resolution of the images I'm counting off of to overlay as a separate signal channel but that didn't seem to work properly. I'd love to help/contribute but my python skills are fairly basic. If you could provide some guidance I could try to help though.

**Comment by janfpl** [comment]
I take back what I said earlier about resizing the downsampled boundaries to the original resolution not working, turns out it was an ImageJ-export related problem.

Now I can go back and adjust the cell markers now with the overlay of the original images and the atlas boundaries. This reminds me of the correspondance point adjustment which is a really nice feature of WholeBrain (R package). It may be useful to use these data to allow for some sort of manual correction of registration error in the future.

**Comment by davoudian** [comment]
was playing around with this, @adamltyson just a heads up that `cellfinder_curate_new ` seems depreciated in the most recent version of cellfinder (0.4.15).

Think it would be worth keeping around for cellfinder

**Comment by adamltyson** [comment]
Yep this is my mistake. The aim is for all of this functionality to end up in the napari plugin (see issue - https://github.com/brainglobe/cellfinder-napari/issues/3), but I haven't had time to move it all.

**Comment by adamltyson** [comment]
@davoudian when you said that `cellfinder_curate_new` was deprecated, did you mean you got this warning from napari?
```
The 'gui_qt()' context manager is deprecated.
```

**Comment by davoudian** [comment]
@adamltyson that's the warning I get when using an older version of cell finder, but the manual curation still seemed to work

**Comment by adamltyson** [comment]
That's a napari deprecation.  I'll sort it out, but it shouldn't affect cellfinder for now.

**Comment by adamltyson** [comment]
Closing this, as the work has been moved to https://github.com/brainglobe/cellfinder-napari/issues/3
